hivemq-edge/src/frontend/src/extensions/datahub/designer/schema/SchemaPanel.tsx

```diff
@@ -1,23 +1,89 @@
 import { datahubRJSFWidgets } from '@datahub/designer/datahubRJSFWidgets.tsx'
 import { MOCK_SCHEMA_SCHEMA } from '@datahub/designer/schema/SchemaData.ts'
+import { getSchemaFamilies } from '@datahub/designer/schema/SchemaNode.utils.ts'
+import useDataHubDraftStore from '@datahub/hooks/useDataHubDraftStore.ts'
 
 export const SchemaPanel: FC<PanelProps> = ({ selectedNode, onFormSubmit }) => {
+  const { data: allSchemas } = useGetAllSchemas()
   const { nodes } = useDataHubDraftStore()
-  // const [fields, setFields] = useState<string[] | null>(null)
-
-  const data = useMemo(() => {
+  const [formData, setFormData] = useState<SchemaData | null>(() => {
     const adapterNode = nodes.find((e) => e.id === selectedNode) as Node<SchemaData> | undefined
-    return adapterNode ? adapterNode.data : null
-  }, [selectedNode, nodes])
+
+    const internalStatus =
+      typeof adapterNode?.data.version === 'number' ? ResourceStatus.LOADED : adapterNode?.data.version
+
+    return adapterNode ? { ...adapterNode.data, internalStatus } : null
+  })
+
+  const onReactFlowSchemaFormChange = useCallback(
+    (changeEvent: IChangeEvent, id?: string | undefined) => {
+      // id have form "root_XXXXXX", which makes the test unsafe
+      if (id?.includes('name')) {
+        const schema = allSchemas?.items?.findLast((e) => e.id === changeEvent.formData.name)
+        if (schema) {
+          setFormData({
+            name: schema.id,
+            type: enumFromStringValue(SchemaType, schema.type) || SchemaType.JSON,
+            version: schema.version || ResourceStatus.MODIFIED,
+            schemaSource: atob(schema.schemaDefinition),
+            internalVersions: getSchemaFamilies(allSchemas?.items || [])[schema.id].versions,
+            internalStatus: ResourceStatus.LOADED,
+          })
+        } else {
+          setFormData({
+            internalStatus: ResourceStatus.DRAFT,
+            name: changeEvent.formData.name,
+            type: SchemaType.JSON,
+            version: ResourceStatus.DRAFT,
+            schemaSource: MOCK_JSONSCHEMA_SCHEMA,
+          })
+        }
+      }
+      if (id?.includes('type') && formData) {
+        if (formData.type !== changeEvent.formData.type) {
+          setFormData({
+            ...formData,
+            type: changeEvent.formData.type,
+            schemaSource: changeEvent.formData.type === SchemaType.JSON ? MOCK_JSONSCHEMA_SCHEMA : MOCK_PROTOBUF_SCHEMA,
+          })
+        }
+      }
+      if (id?.includes('schemaSource') && formData && formData.internalStatus === ResourceStatus.LOADED) {
+        setFormData({
+          ...formData,
+          version: ResourceStatus.MODIFIED,
+          internalStatus: ResourceStatus.MODIFIED,
+        })
+      }
+      if (id?.includes('version') && formData) {
+        const schema = allSchemas?.items?.find(
+          (e) => e.id === formData.name && e.version?.toString() === changeEvent.formData.version
```

Comment:
suggestion: same as above let's use `schema` for the iteration item to be more specific

```suggestion
          (schema) => schema.id === formData.name && e.version?.toString() === changeEvent.formData.version
```